Update doc `dql-custom-walkers.rst` with an output walker to interpolate parameters into SQL (#12060)
doctrine/doctrine2: docs/en/cookbook/dql-custom-walkers.rst

            }
            if ($typeName instanceof ParameterType) {
                return $processParameterType($typeName)($value);
            }
            if ($typeName instanceof ArrayParameterType && is_array($value)) {
                $type = ArrayParameterType::toElementParameterType($typeName);
                return implode(', ', array_map($processParameterType($type), $value));
            }

            throw new ValueNotConvertible($value, $typeName);
        }
    }

Then you may get raw SQL with:

.. code-block:: php

    <?php
    $query
        ->where('t.int IN (:ints)')->setParameter(':ints', [1, 2])
        ->orWhere('t.string IN (?0)')->setParameter(0, ['3', '4'])
        ->orWhere("t.bool = ?1")->setParameter('?1', true)
        ->orWhere("t.string = :string")->setParameter(':string', 'ABC')
        ->setHint(\Doctrine\ORM\Query::HINT_CUSTOM_OUTPUT_WALKER, InterpolateParametersSQLOutputWalker::class)
        ->getSQL();

The where clause of returned SQL should be like:

.. code-block:: sql

    WHERE t0_.int IN (1, 2)
       OR t0_.string IN ('3', '4')
       OR t0_.bool = 1

>        OR t0_.string = 'ABC'

Reviewer: Ok, then what? You should probably state the use case for this explicitly.

Author: This example would help https://stackoverflow.com/questions/2095394/doctrine-how-to-print-out-the-real-sql-not-just-the-prepared-statement and https://stackoverflow.com/questions/4155288/how-to-write-union-in-doctrine-2-0/79647475#79647475

Reviewer: You should probably include a summary of this in this doc you are contributing.

allow you write DQL queries instead of NativeQueries to make use of
vendor specific features.

> Modify the Output Walker to get the raw SQL with interpolated parameters

Reviewer: ```suggestion
Modifying the Output Walker to get the raw SQL with interpolated parameters
```

> --------------------------------------------------------

Reviewer: The last dash should be under the last letter of the title

Author: ```suggestion
---------------------------------------------------------------------------
```

> Sometimes we may want to log or trace the raw SQL being generated from its DQL,

Reviewer: You should elaborate here. I think your need is to get a SQL statement that you can replay yourself, with an SQL client. If so, please say so explicitly and clearly.

Author: For me there's two main usage of raw SQL: log the SQL being executed to debug possible slow queries in future, and UNION results of multiple DQL with WHERE as only DBAL query supports UNION as in https://stackoverflow.com/questions/4155288/how-to-write-union-in-doctrine-2-0/79647475#79647475 and https://github.com/n0099/open-tbm/blob/c960f09242097937403880a6ebd02ef4946ce3a8/be/src/PostsQuery/QueryResult.php#L219-L237.

Author: ```suggestion
Sometimes we may want to log or trace the raw SQL being generated from its DQL
for profiling slow queries in the future or audit queries that changed many rows
```

Reviewer: I would say "afterwards" rather than "in the future", because the future has a tendency to become the past at some point.

Author: ```suggestion
Sometimes we may want to log or trace the raw SQL being generated from its DQL
for profiling slow queries afterwards or audit queries that changed many rows
```

> ``$query->getSQL()`` will give us the prepared statment being passed to database

Reviewer: ```suggestion
``$query->getSQL()`` will give us the prepared statement being passed to database
```

with all values of SQL params being replaced by positional ``?`` or named ``:name``

> as params are innerpolated into prepared statment by the database while executing the SQL.

Reviewer: ```suggestion
as parameters are interpolated into prepared statements by the database while executing the SQL.
```

> ``$query->getParameters()`` will give us details about SQL params that we've provided.

Reviewer: Use parameters instead of params everywhere.

So we can create an output walker to interpolate all SQL params that will be
passed into prepared statement in PHP before database handle them internally:

.. code-block:: php

>     <?php

Reviewer: Please use `literalinclude` for this (look for usage in the current docs, I think you will find an example)
